lib/oidcstrategy.js

```diff
@@ -620,38 +619,62 @@ Strategy.prototype.collectInfoFromReq = function(params, req, next) {
+    // user provided tenantIdOrName will be ignored for redirectUrl, since we saved the one we used in session
+    if (params.tenantIdOrName)
+      log.info(`user provided tenantIdOrName '${params.tenantIdOrName}' is ignored for redirectUrl, we will use the one stored in session`);
+    params.tenantIdOrName = tuple['tenantIdOrName'];
   } else {
     params.policy = req.query.p ? req.query.p.toLowerCase() : null;
   }
 
+  // if we are not using the common endpoint, but we have tenantIdOrName, just ignore it
+  if (!self._options.isCommonEndpoint && params.tenantIdOrName) {
+    log.info(`identityMetadata is tenant-specific, so we ignore the tenantIdOrName '${params.tenantIdOrName}'`);
+    params.tenantIdOrName = null;
+  }
+
+  // if we are using the common endpoint and we want to validate issuer, then user has to 
+  // provide issuer in config, or provide tenant id or name using tenantIdOrName option in
+  // passport.authenticate. Otherwise we won't know the issuer.
+  if (self._options.isCommonEndpoint && self._options.validateIssuer &&
+    (!self._options.issuer && !params.tenantIdOrName))
+    return next(new Error('In collectInfoFromReq: issuer or tenantIdOrName must be provided in order to validate issuer on common endpoint'));
+
   // for B2C, we must have policy
   if (self._options.isB2C && !params.policy)
-    return self.failWithLog('In collectInfoFromReq: policy is missing');
+    return next(new Error('In collectInfoFromReq: policy is missing'));
+  // for B2C, if we are using common endpoint, we must have tenantIdOrName provided
+  if (self._options.isB2C && self._options.isCommonEndpoint && !params.tenantIdOrName)
+    return next(new Error('In collectInfoFromReq: we are using common endpoint for B2C but tenantIdOrName is not provided'));
 
   // -------------------------------------------------------------------------
-  // create a cachekey and an Metadata object instance  
+  // calculate metadataUrl, create a cachekey and an Metadata object instance  
   // we will fetch the metadata, save it into the object using the cachekey       
-  // -------------------------------------------------------------------------        
-  var metadataUrl = self._options.identityMetadata;
+  // -------------------------------------------------------------------------   
+  var metadataUrl = self._options.identityMetadata;   
 
-  if (self._options.isB2C) {
-    metadataUrl = metadataUrl.concat(`&p=${params.policy}`);
-    params.cachekey = 'policy: ' + params.policy;  // B2C metadata cachekey
-    log.info(`B2C metadataUrl is: ${metadataUrl}`);
-  } else {
-    params.cachekey = 'ordinary'; // non B2C metadata cachekey
-    log.info(`metadataUrl is: ${metadataUrl}`);
+  // if we are using common endpoint and we are given the tenantIdOrName, let's replace it
+  if (self._options.isCommonEndpoint && params.tenantIdOrName) {
+    metadataUrl = metadataUrl.replace('/common/', `/${params.tenantIdOrName}/`);
```

Comment:
I understand that's where it will fail if we don't validate. I'm suggesting we should validate before then so we can give a better error message if it's set to something invalid. This is true for both tenantIdOrName and the policy name.